### PR TITLE
feat: add server support for kafka

### PR DESCRIPTION
**Pull request by amrutprabhu**
**Description**

This is a pull request for adding support to utilize the servers section to create kafka brokers config in the properties file. 

**Related issue(s)**
https://github.com/asyncapi/java-spring-cloud-stream-template/issues/118

### CODE REVIEW

**Comment by amrutprabhu** [comment]
**Note** : I was not able to test this. I don't have much experience in coding in javascript. Please comment on the code and also could you tell me how to test the changes?

**Comment by amrutprabhu** [comment]
**Note** : I was not able to test this. I don't have much experience in coding in javascript. Please comment on the code and also could you tell me how to test the changes?

File: filters/all.js
Comment (derberg) [inline filters/all.js:555]:
```suggestion
       brokers += `${url},`;
```

File: filters/all.js
Comment (derberg) [inline filters/all.js:551]: you expect that someone uses `port` word to specify port + that user always provides a default. This if fine, but needs to be documented in the readme.

**Review by derberg — COMMENTED** [review]
@amrutprabhu looks good so far, just left few comments. To test, best just run generation. Install AsyncAPI generator first and then:

`ag yourtestasyncapifule.yml ./ -o output --force-write -p useServers=true -p otherparams=something`

run this command from the root of the repo with the template, `./` tells generator to pick up the template from where you run it. Generated files will end up in `output` folder. the `--force-write` flag is to ignore warnings that you can override some of your uncommited changes - you are secure as you provide the output dir

File: filters/all.js
Comment (damaru-inc) [reply filters/all.js:145]: Yeah, line 145 looks wrong, it should be assigned brokerSettings not additionalSubs right?

File: README.md
Comment (amrutprabhu) [reply README.md]: Updated this.

File: filters/all.js
Comment (amrutprabhu) [reply filters/all.js]: Ya.. That's my bad.. Updated it now.

File: filters/all.js
Comment (amrutprabhu) [reply filters/all.js]: Updated this also.

File: filters/all.js
Comment (amrutprabhu) [reply filters/all.js:551]: So I was following the example from here: https://raw.githubusercontent.com/asyncapi/asyncapi/2.0.0/examples/2.0.0/streetlights.yml
and I also saw that the a logic was being used in the nodejs project. at this line : https://github.com/asyncapi/nodejs-template/blob/fff91aa1f76db4682332f16bb91cdd78fb536ceb/template/config/common.yml#L41

I was trying to do something like [this](https://github.com/asyncapi/nodejs-template/blob/fff91aa1f76db4682332f16bb91cdd78fb536ceb/filters/all.js#L173) function does, But the function was too big. So I thought `port` was one of the predefined variables in the spec, but I was wrong. 

I could omit this section to support variables and Just keep it simple. 

What you do think?

Comment (derberg) [reply filters/all.js:551]: I'm the author of this beast function 😅 

spec does not specify any convention for variable names. Why can't you take the entire function for replacing variables? In the end you just want to have a "working" URL without variables in it right? We can extract this function from nodejs template to `generator-filters` repo where we have reusable helpers for templates

File: filters/all.js
Comment (derberg) [inline filters/all.js:543]: what `ret` stands for?

File: filters/all.js
Comment (amrutprabhu) [reply filters/all.js:551]: So should I just copy the function for now and you would extract this to a `generator-filter` repo?

File: filters/all.js
Comment (amrutprabhu) [reply filters/all.js:543]: It was for returning a value.. ill refactor this to a proper name.

File: filters/all.js
Comment (amrutprabhu) [reply filters/all.js:551]: @derberg .. what you do think about the suggestion I had made?

File: filters/all.js
Comment (MichaelDavisSolace) [reply filters/all.js]: I'm trying to run this. I grabbed a copy of your fork, Amrut, and switched to this branch. When I ran it I got an error saying that it didn't accept the useServers parameter, even though I see it clearly in your package.json. I assume this works for you?

**Comment by damaru-inc** [comment]
I'm trying to run this. I grabbed a copy of your fork, Amrut, and switched to this branch. When I ran it I got an error saying that it didn't accept the useServers parameter, even though I see it clearly in your package.json. I assume this works for you?

**Comment by amrutprabhu** [comment]
@damaru-inc I haven't tested the changes, as I needed some input to go forward in this comment: https://github.com/asyncapi/java-spring-cloud-stream-template/pull/125#discussion_r667178497

File: filters/all.js
Comment (derberg) [reply filters/all.js:551]: sorry, holidays. The best would be if you could perform the extraction. Not exactly extraction, but just add the same filter to `generator-filters` and then use here (but first make sure it works for you of course. I can later refactor NodeJS template to use this generic filter too

Comment (derberg) [reply filters/all.js:551]: @amrutprabhu do you need some support from my side?

Comment (amrutprabhu) [reply filters/all.js]: Hey @derberg ,
So I have updated the code and tested this by right now copying the function `replaceVariablesWithValues` from the `nodejs-template`.

I have also raised a PR in the `generator-filter` repo to have this function as the common function. Here is the PR for it. 
https://github.com/asyncapi/generator-filters/pull/46

Can you have a look at this ?

File: filters/all.js
Comment (amrutprabhu) [inline filters/all.js:566]: Will remove this function once this PR is solved: https://github.com/asyncapi/generator-filters/pull/46

File: filters/all.js
Comment (amrutprabhu) [reply filters/all.js]: Once the `generator-filters` PR is solved, Do I call the function name directly?

File: filters/all.js
Comment (amrutprabhu) [reply filters/all.js:565]: There is also a code smell in this function, which will be fixed once this function is removed.

**Comment by amrutprabhu** [comment]
@derberg 
So I have updated the code accordingly. 
Can you review the PR https://github.com/asyncapi/generator-filters/pull/46 in order to proceed further?

File: filters/all.js
Comment (amrutprabhu) [reply filters/all.js]: @derberg I have fixed this

**Comment by amrutprabhu** [comment]
@derberg Can you have a look at the changes?
I have updated the code wot use the generator filter from https://github.com/asyncapi/generator-filters/pull/46

**Review by derberg — COMMENTED** [review]
@amrutprabhu please have a look at linter issues, they added a lot of noise to the PR and make review harder too

**Comment by amrutprabhu** [comment]
Hey @derberg ,

I was waiting for the pipeline to complete.. But apparently its still waiting...

**Comment by derberg** [comment]
@amrutprabhu there are some issues with code reported by linter. You can run liner on your own locally and not wait for CI to report errors: `npm run lint`.

Ci is blocked and need manual accept from me or other maintainers as it is your first-time contribution to this repo and yeah....GH requires manual approval on CI jobs because of damn bitcoin miners that abuse free CI services 😞 

code wise all good, did you test it manually? Please also update to matest master.

**Comment by amrutprabhu** [comment]
@derberg . So I Have tested the change locally and also fixed the linting issues. 

I have used the commands to test my changes and it works as expected.
```
ag asyncapi.yml ./ -o sample-project  -p javaPackage=com.test.producer -p view=provider --force-write

and 
ag asyncapi.yml ./ -o sample-project  -p javaPackage=com.test.producer -p view=provider --force-write -p useServers=true
```
asyncapi.yml 
```
asyncapi: 2.0.0
info:
  title: Banking backend events
  version: 1.0.0
servers:
  development:
    url: a-nuc:{port}
    protocol: kafka
    variables:
      port:
        default: '9092'
        enum:
          - '8080'
          - '8883'
channels:
  banking.transaction.000:
    description: |-
      This contains events related to transactions.
    publish:
      operationId: consumeTransactionEvent
      message:
        $ref: '#/components/messages/transactionEvent'
      bindings:
        kafka:
          groupId: myGroupId

components:
  messages:
    transactionEvent:
      name: trasactionEvent
      contentType: 'application/json'
      payload:
        $ref: '#/components/schemas/TransactionEventPayload'
  schemas:
    TransactionEventPayload:
      type: object
      properties:
        transactionId:
          description: Transaction Id
          type: string
        amount:
          type: integer
```

**Review by derberg — APPROVED** [review]
Thanks for the contribution. @MichaelDavisSolace need another approve in my opinion as it is adding something new that you and your use cases did not need

**Review by damaru-inc — APPROVED** [review]

**Comment by damaru-inc** [comment]
Thanks for this!

**Comment by derberg** [comment]
@damaru-inc that was quick! thanks a lot! and thanks a lot @amrutprabhu for your patience and a very good contribution 🚀

**Comment by amrutprabhu** [comment]
Cool.
Thanks, @derberg, and @damaru-inc for this. :tada: :tada: 
I got to learn some javascript :smile:

**Pull request merged**
